### ISSUE TITLE
Support non-`\` trigger characters

### ISSUE THREAD

**Issue by yzhang-gh** [yzhang-gh/vscode-markdown#331]
1. Type `\` -> autocomplete panel shows automatically.
2. `esc` -> abort/hide panel
3. `ctrl-space` -> autocomplete works
4. Type `a` -> autocomplete shows `\And`, ...
5. `esc` -> abort/hide panel
6. `ctrl-space` (cursor after `\a`) -> __only basic text completes__

_Originally posted by @hannesfrank in https://github.com/neilsustc/vscode-markdown/issues/326#issuecomment-441386130_

**Comment by hannesfrank** [comment]
I think I could implement this. I'm just brainstorming a few implementation details here:

Let

- `m = length of suggestion prefix` (i.e. 3 for `before \abc| after`). As usual, we get this prefix by a regex like `\\([^\\])*$` from `beforeText`.
- `n = number of suggestions` (this is constant and about a few 100s here)

The __naive algorithm__ would match to iterate over all suggestions and check if it starts with the prefix, time `O(n * m)`.

A __nice algorithm__ would use a prefix tree built of all the suggestions with the paths representing a prefix and all the suggestions on the leaves (in sorted order). Searching a prefix leads to a node `v` (`O(m)`) and we return all suggestions of the subtree with root `v`. There are a few ways to pick them up, either by naively traversing the tree or with the help of a few extra pointers:

- for each inner node hold a pointer to the leftmost node of its subtree and
- for each leaf hold a pointer to the next leaf on the right.

The prefixtree could be precomputed and stored in the code as an object like
```javascript
const prefix_tree = {
  "a": { "n": { 
    "d": { "suggestion": "And", "arguments": 0, "description": "Ampersand symbol &" },
    "g": { "suggestion": "angle", "arguments": 0, "description": "..." }
  } ... }
}
```
The time for this is linear in the number of returned suggestions, but the tree traversal might be slow, especially when you just type one letter, and linking the tree properly must be done at plugin initialization because it is impossible using object literals.

---

I think the naive algorithm wins because it has way lower code complexity and might even be faster with this few suggestions.

Thoughts?

**Comment by neilsustc** [comment]
Thanks for the thoughts!

I also thought for a while last night but didn't figure out a good solution.

---

The most simple one, as you said, **the naive algorithm**, is easy to implement. I checked the code of [LaTeX Workshop](https://github.com/James-Yu/LaTeX-Workshop). They seemed to be using this solution but I was a little confused by their code.

The disadvantages are

1. need to calculate completions on every input, which could result in bad performance (?)

   but maybe not that bad as we expected?

2. I suspect whether we can still leverage the "fuzzy match" features

   ![fuzzy match](https://user-images.githubusercontent.com/7588612/49016494-2ee91280-f17e-11e8-9504-f75dcf41087e.png)

---

As for **the prefixtree one**, smart idea to solve problem 1 although might be an overkill for our case. I agree with you that "the naive algorithm wins because it has way lower code complexity and might even be faster with this few suggestions".

---

What do you think of the problem 2 "fuzzy match"?

**Comment by hannesfrank** [comment]
tldr: see below `hr` for actually useful ideas :smile:

I discovered 2 problems with my ideas (that's why I had to delete that comment, but if we use that as a base of discussions, I might forced to put these flawed thoughts online again :disappointed:? ):

1. I made gathering the suggestions from the prefix tree way to complex: It is smarter to hold a sorted array of all suggestions and each node of the tree just links to the starting (and end) index of the subsequence of suggestions represented by the subtree.
2. As you mentioned: I did not even think about fuzzy matching. A prefix tree only does prefix matching. I'm sure that there must be some clever data strucutres (which I am looking forward to research later) because you fuzzy search a directory with a million files almost instantly.

---

But as I continue researching VSCode plugins - __plot twist__ - it turns out that (unlike I intuitively expected) the `provideCompletionItems` gets called exactly once when you trigger the completion (either automatically or with `ctrl+space`) and _not with every keypress_ which means the fuzzy matching is done automatically by VSCode as you type.

This means you just have to return
- _all_ math completions when you detect you are in a math environment (or as we currently just expect to be in one when someone types `\`) and
- _all_ image paths in the workspace when you detect you are completing the path of an image.

This also seems to be exactly what LaTeX-Workshop is doing. The completion provider asks every sub completer to return their list and e.g. the `commands` provider just returns _all_ commands - the default ones and the ones from imported packages.

**Comment by neilsustc** [comment]
> It turns out that (unlike I intuitively expected) the `provideCompletionItems` gets called exactly once when you trigger the completion (either automatically or with `ctrl+space`) and not with every keypress which means the fuzzy matching is done automatically by VSCode as you type.

That's true. I noticed it and edited my comment just after it is posted

![image](https://user-images.githubusercontent.com/7588612/49022654-44196d80-f18d-11e8-8189-051752613657.png)

> This means you just have to return
> - _all_ math completions when you detect you are in a math environment (or as we currently just expect to be in one when someone types `\`) and
> - _all_ image paths in the workspace when you detect you are completing the path of an image.

Yes, then we need to also add `[a-zA-Z]` as trigger characters. Here comes the problem, the `CompletionItem`s should not be the same. Otherwise `\an` accepting the completion `\and` will become `\anand`. We need to change the `range` property of the `CompletionItem`s.

**Comment by neilsustc** [comment]
You don't need to delete your post if you think it is wrong. Adding a warning or using ~~strikethrough~~ (`~~`) might be good enough 😊

**Comment by hannesfrank** [comment]
I would leave `\` as the only trigger character and require the user to press `ctrl+space` when they want to get completions again. 

The default `CompletionItem.range` (replacng the whole word in front of the cursor) should also be fine for math completions as `\` marks the beginning of a new word (see edit above)?

**Comment by neilsustc** [comment]
Looks reasonable. Would you be willing to open a PR for it?

**Comment by hannesfrank** [comment]
Yes, I'm working on it.
